File: drive/scripts/research_data_mcp/candidate_key.py
#!/usr/bin/env python3
"""Canonical Discover candidate identity (D0b — mirrors frontend candidateKey.js).

Precedence:
1. server-provided candidate_key
2. dataset_id
3. canonical DOI
4. source-specific external identifier
5. canonical URL
6. namespaced normalized title fallback (last resort; never raw title alone)

Typed prefixes prevent cross-type collisions:
  dataset:<id> | doi:<doi> | source:<provider>:<id> | url:<url> | title:<provider>:<title>
"""

from __future__ import annotations

import re
from typing import Any
from urllib.parse import urlsplit, urlunsplit


def _trim(value: Any) -> str:
    return str(value or "").strip()


def canonicalize_doi(value: Any) -> str:
    doi = _trim(value)
    if not doi:
        return ""
    doi = re.sub(r"^doi:\s*", "", doi, flags=re.IGNORECASE)
    doi = re.sub(r"^https?://(dx\.)?doi\.org/", "", doi, flags=re.IGNORECASE)
    return doi.lower()


def canonicalize_url(value: Any) -> str:
    """Canonical URL for identity — keep meaningful query params; drop fragment."""
    raw = _trim(value)
    if not raw:
        return ""
    try:
        parts = urlsplit(raw)
        if not parts.scheme or not parts.hostname:
            return raw.lower()
        scheme = parts.scheme.lower()
        host = parts.hostname.lower()
        port = parts.port
        if port is not None and not (
            (scheme == "http" and port == 80) or (scheme == "https" and port == 443)
        ):
            netloc = f"{host}:{port}"
        else:
            netloc = host
        if parts.username:
            auth = parts.username
            if parts.password:
                auth = f"{auth}:{parts.password}"
            netloc = f"{auth}@{netloc}"
        path = parts.path or ""
        if path != "/" and path.endswith("/"):
            path = path.rstrip("/")
        return urlunsplit((scheme, netloc, path, parts.query, ""))
    except Exception:
        return raw.lower()


def normalize_title(value: Any) -> str:
    return re.sub(r"\s+", " ", _trim(value)).lower()


def slugify_provider(value: Any) -> str:
    """Unicode-safe provider namespace (D0.2).

    NFKC → trim → casefold → keep letters/numbers/._- → collapse other runs to _.
    """
    import unicodedata

    raw = _trim(value)
    if not raw:
        return "unknown"
    s = unicodedata.normalize("NFKC", raw).casefold().strip()
    parts: list[str] = []
    prev_us = False
    for ch in s:
        if ch.isalnum() or ch in "._-":
            parts.append(ch)
            prev_us = False
        else:
            if not prev_us:
                parts.append("_")
                prev_us = True
    slug = "".join(parts).strip("_")
    slug = "".join(list(slug)[:80])
    if not any(ch.isalnum() for ch in slug):
        return "unknown"
    return slug or "unknown"


def _provider_slug(row: dict[str, Any]) -> str:
    host = ""
    url = _trim(row.get("url") or row.get("source_url") or row.get("resolved_url") or "")
    if url:
        try:
            host = (urlsplit(url).hostname or "").lower().removeprefix("www.")
        except Exception:
            host = ""
    raw = (
        _trim(row.get("provider"))
        or _trim(row.get("publisher"))
        or _trim(row.get("source"))
        or _trim(row.get("collect_via"))
        or _trim(row.get("kind"))
        or host
        or ""
    )
    return slugify_provider(raw)


def _source_external_id(row: dict[str, Any]) -> tuple[str, str] | None:
    kind = _trim(row.get("kind")).lower()
    handle = _trim(row.get("handle") or row.get("open_handle"))
    if handle.startswith("hf:"):
        return ("huggingface", handle[3:])
    if handle.startswith("doi:"):
        return None
    if kind == "huggingface":
        ext = _trim(row.get("hf_id") or row.get("id") or row.get("external_id"))
        if ext and "://" not in ext:
            return ("huggingface", ext)
    external = _trim(row.get("external_id") or row.get("source_id"))
    if external:
        return (_provider_slug(row), external)
    return None


def candidate_key(row: dict[str, Any] | None) -> str:
    """Return typed candidate key, or empty string if nothing usable."""
    if not isinstance(row, dict):
        return ""

    server_key = _trim(row.get("candidate_key"))
    if server_key:
        return server_key

    dataset_id = _trim(row.get("dataset_id"))
    if not dataset_id and _trim(row.get("kind")).lower() == "local_registry":
        dataset_id = _trim(row.get("id"))
    if dataset_id:
        return f"dataset:{dataset_id}"

    doi = canonicalize_doi(row.get("doi"))
    if doi:
        return f"doi:{doi}"

    ext = _source_external_id(row)
    if ext and ext[1]:
        return f"source:{ext[0]}:{ext[1]}"

    url = canonicalize_url(row.get("resolved_url") or row.get("source_url") or row.get("url"))
    if url:
        return f"url:{url}"

    title = normalize_title(row.get("title") or row.get("name"))
    if title:
        return f"title:{_provider_slug(row)}:{title}"

    return ""


def with_candidate_key(row: dict[str, Any] | None) -> dict[str, Any] | None:
    """Return a shallow copy stamped with candidate_key (does not invent empty keys)."""
    if not isinstance(row, dict):
        return row
    key = candidate_key(row)
    if not key or row.get("candidate_key") == key:
        return row
    out = dict(row)
    out["candidate_key"] = key
    return out


def stamp_rows(rows: list[Any] | None) -> list[Any]:
    if not rows:
        return []
    stamped: list[Any] = []
    for row in rows:
        stamped.append(with_candidate_key(row) if isinstance(row, dict) else row)
    return stamped

File: drive/src/v2/candidateKey.js
/**
 * Canonical Discover candidate identity (D0a).
 *
 * Precedence:
 * 1. server-provided candidate_key
 * 2. dataset_id
 * 3. canonical DOI
 * 4. source-specific external identifier
 * 5. canonical URL
 * 6. namespaced normalized title fallback
 *
 * Typed prefixes prevent cross-type collisions.
 */

function trim(value) {
  return String(value ?? "").trim();
}

/** Canonical DOI: strip resolver/prefix, lowercase. */
export function canonicalizeDoi(value) {
  let doi = trim(value);
  if (!doi) return "";
  doi = doi.replace(/^doi:\s*/i, "");
  doi = doi.replace(/^https?:\/\/(dx\.)?doi\.org\//i, "");
  return doi.toLowerCase();
}

/**
 * Canonical URL for identity:
 * - prefer backend-resolved URL when present
 * - trim, drop fragment
 * - lowercase scheme + hostname
 * - remove default ports
 * - keep meaningful query parameters
 */
export function canonicalizeUrl(value) {
  const raw = trim(value);
  if (!raw) return "";
  try {
    const u = new URL(raw);
    u.hash = "";
    u.protocol = u.protocol.toLowerCase();
    u.hostname = u.hostname.toLowerCase();
    if (
      (u.protocol === "http:" && u.port === "80") ||
      (u.protocol === "https:" && u.port === "443")
    ) {
      u.port = "";
    }
    // Normalize pathname trailing slash only for root
    if (u.pathname !== "/" && u.pathname.endsWith("/")) {
      u.pathname = u.pathname.replace(/\/+$/, "");
    }
    return u.toString();
  } catch {
    return raw.toLowerCase();
  }
}

export function normalizeTitle(value) {
  return trim(value).replace(/\s+/g, " ").toLowerCase();
}

/**
 * Canonical action URL for probe / collect / Add-to-lab.
 * Same field preference as the url: identity tier (before DOI fallback):
 * resolved_url → source_url → url → DOI resolver.
 */
export function discoverCandidateUrl(row) {
  if (!row || typeof row !== "object") return "";
  const raw = trim(row.resolved_url || row.source_url || row.url);
  if (raw) return raw;
  const doi = canonicalizeDoi(row.doi);
  if (doi) return `https://doi.org/${doi}`;
  const handle = trim(row.open_handle || row.handle);
  if (handle.startsWith("doi:")) return `https://doi.org/${canonicalizeDoi(handle.slice(4))}`;
  return "";
}

/**
 * Unicode-safe provider namespace (D0.2).
 * NFKC → trim → lowercase → keep letters/numbers/._- → collapse other runs to _.
 */
export function slugifyProvider(value) {
  const raw = trim(value);
  if (!raw) return "unknown";
  let s = raw.normalize("NFKC").trim().toLowerCase();
  s = s.replace(/[^\p{L}\p{N}._-]+/gu, "_").replace(/^_+|_+$/gu, "");
  s = Array.from(s).slice(0, 80).join("");
  if (!/[\p{L}\p{N}]/u.test(s)) return "unknown";
  return s || "unknown";
}

function providerSlug(row) {
  const host = (() => {
    const url = trim(row?.url || row?.source_url || row?.resolved_url || "");
    if (!url) return "";
    try {
      return new URL(url).hostname.replace(/^www\./, "").toLowerCase();
    } catch {
      return "";
    }
  })();
  const raw =
    trim(row?.provider) ||
    trim(row?.publisher) ||
    trim(row?.source) ||
    trim(row?.collect_via) ||
    trim(row?.kind) ||
    host ||
    "";
  return slugifyProvider(raw);
}

function sourceExternalId(row) {
  const kind = trim(row?.kind).toLowerCase();
  const handle = trim(row?.handle || row?.open_handle);
  if (handle.startsWith("hf:")) return { provider: "huggingface", id: handle.slice(3) };
  if (handle.startsWith("doi:")) return null; // DOI handled separately
  if (kind === "huggingface") {
    const id = trim(row?.hf_id || row?.id || row?.external_id);
    if (id && !id.includes("://")) return { provider: "huggingface", id };
  }
  const external = trim(row?.external_id || row?.source_id);
  if (external) return { provider: providerSlug(row), id: external };
  return null;
}

/**
 * @param {object|null|undefined} row
 * @returns {string} typed candidate key, or "" if empty
 */
export function candidateKey(row) {
  if (!row || typeof row !== "object") return "";

  const serverKey = trim(row.candidate_key);
  if (serverKey) return serverKey;

  const datasetId = trim(row.dataset_id || (row.kind === "local_registry" ? row.id : ""));
  if (datasetId) return `dataset:${datasetId}`;

  const doi = canonicalizeDoi(row.doi);
  if (doi) return `doi:${doi}`;

  const ext = sourceExternalId(row);
  if (ext?.id) return `source:${ext.provider}:${ext.id}`;

  const url = canonicalizeUrl(row.resolved_url || row.source_url || row.url);
  if (url) return `url:${url}`;

  const title = normalizeTitle(row.title || row.name);
  if (title) return `title:${providerSlug(row)}:${title}`;

  return "";
}

/** Stamp candidate_key onto a row copy (does not mutate). */
export function withCandidateKey(row) {
  if (!row || typeof row !== "object") return row;
  const key = candidateKey(row);
  if (!key || row.candidate_key === key) return row;
  return { ...row, candidate_key: key };
}

/**
 * Exact job↔candidate association only (no title matching).
 * Jobs may carry candidate_key / connector_id at top level or under request/plan.
 */
export function jobMatchesCandidate(job, row) {
  if (!job || !row) return false;
  const rowKey = candidateKey(row);
  const jobKey = trim(
    job.candidate_key ||
      job.request?.candidate_key ||
      job.plan?.candidate_key ||
      "",
  );
  if (rowKey && jobKey && rowKey === jobKey) return true;

  const rowConnector = trim(
    row.connector_id || row.probe_connector_id || row.connector?.connector_id || row.connector?.id || "",
  );
  const jobConnector = trim(
    job.connector_id ||
      job.request?.connector_id ||
      job.plan?.connector_id ||
      "",
  );
  if (rowConnector && jobConnector && rowConnector === jobConnector) return true;
  return false;
}

export function isCandidateQueued(row, jobs = []) {
  if (!row || !Array.isArray(jobs) || !jobs.length) return false;
  if (row.queued === true && (row.candidate_key || row.connector_id)) {
    // Trust only if already stamped from an exact association path
    return jobs.some((j) => jobMatchesCandidate(j, row));
  }
  return jobs.some((j) => {
    const status = String(j?.status || "").toLowerCase();
    if (!["pending_approval", "queued", "running"].includes(status)) return false;
    return jobMatchesCandidate(j, row);
  });
}

/** A row is the selected candidate only when something is selected and it names that row. */
export function isSelectedCandidate(selectedId, row) {
  if (!selectedId || !row) return false;
  return selectedId === candidateKey(row) || (Boolean(row.dataset_id) && selectedId === row.dataset_id);
}
